### DISCUSSION TITLE
Weights initialization of the rf neurons

### DISCUSSION THREAD

**Discussion by zeinebCh** [lava-nc/lava-dl#198] [General]
Hello everyone :) 
I'm currently facing difficulties in initializing the weights of the RF-neurons. My objective is to replicate the FFT behavior observed on a white noise signal, which spans frequencies between 10KHz and 20KHz (sampling frequency= 100KHz). Additionally, I have introduced a fundamental frequency component at 15KHz.
When performing the FFT on my signal, I get a peak at the fundamental frequency (here 15 KHz). However when feeding the same signal to my 100 RF-neurons with resonating frequencies covering the whole Bandwidth (10KHz,20KHz / period=[5,10]), I keep getting random peaks.
I also expanded the signal to third dimension (time dimension).

My questions are:
1- Are the neuron weights randomized in lava-dl?
2- How should I Initialize the neuron weights to mimic the FFT.

flollowing is the code I used:
        import numpy as np
        import matplotlib.pyplot as plt
        from lava.lib.dl import slayer
        import torch
        
        # Define the sampling rate and duration
        fs = 100000  # 100 samples in 1ms
        duration = 0.001  # 1ms
        # Generate white noise signal
        noise = np.random.randn(int(fs * duration))
        # Generate sinusoidal signal at 15KHz
        f1 = 15000
        t = np.arange(0, duration, 1 / fs)
        s1 = 2.2 * np.sin(2 * np.pi * f1 * t)
        # Combine the noise and sinusoidal signals
        y = noise + s1
        # Compute the FFT of the time-domain signal
        fft = np.abs(np.fft.fft(y)) / len(y)
        freqs = np.fft.fftfreq(len(y), 1 / fs)
        
        
        input = torch.from_numpy(y).reshape((1, 1, 100)).float()
        input = input.expand(1, 100, 100)
        # Define your weight scaling factor based on the maximum FFT magnitude
        weight_scale = np.max(fft)
        # Initialize the pre_hook_fx function to set weights proportional to the FFT magnitude
        pre_hook_fx = lambda weights: weights * torch.from_numpy(fft).float() / weight_scale
        rf_params = dict(threshold=10, period=[5, 10], decay=0.734, shared_param=False, log_init=True, 
        persistent_state=True)
        
        block = slayer.block.rf.Dense(neuron_params=rf_params, in_neurons=100, out_neurons=100, 
        pre_hook_fx=pre_hook_fx, delay_shift=False, count_log=False)
        out = block(input)
        
I get the following plot:
![fft_vs_rf](https://github.com/lava-nc/lava-dl/assets/86193477/68670bc0-bce0-4e30-a517-7660236b8fec)

I will be extremely thankful to get some help to solve this. My thesis is based on this and it will be a major help for me. Thank you

**Comment by bamsumit** [comment]
@zeinebCh 

> 1- Are the neuron weights randomized in lava-dl?
Yes

> 2- How should I Initialize the neuron weights to mimic the FFT.
You can overwrite the initialization by accessing the weight data tensor `block.synapse.real.weight.data = <torch tensor you like>`

This piece of code might be useful:
```python
import numpy as np
import matplotlib.pyplot as plt
from lava.lib.dl import slayer
import torch
from matplotlib import pyplot as plt

fs = 100e3  # 100 KHz
fc = 15e3   # 15 KHz
sinusoid = 2.2 * np.sin(2 * np.pi * fc / fs * np.arange(2500))
noise = np.random.randn(sinusoid.shape[0])

input = torch.FloatTensor(sinusoid + noise).reshape([1, 1, -1])
rf_params = dict(threshold=2,
                 period= [5, 50],
                 decay=0.01,
                 shared_param=False,
                 log_init=True)
block = slayer.block.rf.Dense(neuron_params=rf_params, in_neurons=1, out_neurons=100)
# Overwrite the weight initialization values
block.synapse.real.weight.data = 0.1 * torch.ones_like(block.synapse.real.weight.data)
block.synapse.imag.weight.data = 0.0 * torch.ones_like(block.synapse.imag.weight.data)
out = block(input)

# Plot spiking events of RF population
ev = np.argwhere(out[0].cpu().data.numpy() > 0)
fig, ax = plt.subplots(1, 2, figsize=(10, 5), sharey=True)
ax[0].plot(ev[:, 1], ev[:, 0], '.', label='spikes')
ax[0].axis([0, input.shape[-1], 0, out.shape[1]])
ax[1].plot(block.neuron.frequency * fs, np.arange(out.shape[1]))
ax[1].vlines(x=fc, ymin=0, ymax=1000, color='gray', linestyle='dotted')
ax[0].set_xlabel('time-steps')
ax[0].set_ylabel('Neuron ID')
ax[0].legend()
ax[1].set_xlabel('Neuron center frequency (Hz)')
```
![image](https://github.com/lava-nc/lava-dl/assets/29907126/8f47524e-abe9-48e0-952c-992df2775d67)

You can play around with frequency of input, and investigate graded spike. That might give more insight on the peak.

__Note:__ A population of RF neurons act like STFT, not FFT.

**Reply by zeinebCh** [reply]
@bamsumit I am extremely thankful to your support. It really helped me.
